source/extensions/common/aws/utility.cc

```diff
@@ -58,32 +67,127 @@ Utility::canonicalizeHeaders(const Http::RequestHeaderMap& headers) {
   formatted_headers.reserve(canonical_headers.size());
   for (const auto& header : canonical_headers) {
     formatted_headers.emplace_back(fmt::format("{}:{}", header.first, header.second));
     parts.emplace_back(formatted_headers.back());
   }
   // need an extra blank space after the canonical headers
-  parts.emplace_back("");
+  parts.emplace_back(EMPTY_STRING);
   const auto signed_headers = Utility::joinCanonicalHeaderNames(canonical_headers);
   parts.emplace_back(signed_headers);
   parts.emplace_back(content_hash);
   return absl::StrJoin(parts, "\n");
 }
 
+std::string Utility::canonicalizePathString(absl::string_view path_string,
+                                            absl::string_view service_name) {
+  // If service is S3, do not normalize but only encode the path
+  if (absl::AsciiStrToLower(service_name) == S3_SERVICE_NAME) {
+    return encodePathSegment(path_string, service_name);
+  }
+  // If service is not S3, normalize and encode the path
+  const auto path_segments = StringUtil::splitToken(path_string, std::string{PATH_SPLITTER});
+  std::vector<std::string> path_list;
+  for (const auto& path_segment : path_segments) {
+    if (path_segment.empty()) {
+      continue;
+    }
+    path_list.emplace_back(encodePathSegment(path_segment, service_name));
+  }
+  auto canonical_path_string =
+      fmt::format("{}{}", PATH_SPLITTER, absl::StrJoin(path_list, std::string{PATH_SPLITTER}));
+  // Handle corner case when path ends with '/'
+  if (!path_string.empty() && path_string.back() == PATH_SPLITTER &&
+      canonical_path_string.size() > 1) {
+    canonical_path_string.push_back(PATH_SPLITTER);
+  }
+  return canonical_path_string;
+}
+
+std::string Utility::encodePathSegment(absl::string_view decoded, absl::string_view service_name) {
+  std::string encoded;
+  for (char c : decoded) {
+    if (std::isalnum(c) || c == '-' || c == '.' || c == '_' || c == '~') {
+      // Escape unreserved chars from RFC 3986
+      encoded.push_back(c);
+    } else if (absl::AsciiStrToLower(service_name) == S3_SERVICE_NAME) {
+      // Do not encode '/' for S3
+      if (c == PATH_SPLITTER) {
+        encoded.push_back(c);
+      } else {
+        // S3 path should be single encoded
+        absl::StrAppend(&encoded, fmt::format(URI_ENCODE, c));
+      }
+    } else {
+      // All other services' path should be double encoded
+      absl::StrAppend(&encoded, fmt::format(URI_DOUBLE_ENCODE, c));
+    }
+  }
+  return encoded;
+}
+
+std::string Utility::canonicalizeQueryString(absl::string_view query_string) {
+  // Sort query string based on param name and append "=" if value is missing
+  const auto query_fragments = StringUtil::splitToken(query_string, QUERY_SEPERATOR);
+  std::vector<std::pair<std::string, std::string>> query_list;
+  for (const auto& query_fragment : query_fragments) {
+    // Only split at the first "=" and encode the rest
+    size_t found = query_fragment.find(QUERY_PARAM_SEPERATOR);
```

Comment:
You should use `absl::StrSplit(query_fragment, absl::MaxSplits('=', 1))` instead.